vllm/core/block_manager_v2.py

```diff
@@ -227,20 +228,69 @@ def fork(self, parent_seq: Sequence, child_seq: Sequence) -> None:
 
     def can_swap_in(self, seq_group: SequenceGroup,
                     num_lookahead_slots: int) -> bool:
-        return False
-
-    def swap_in(self, seq_group: SequenceGroup,
-                num_lookahead_slots: int) -> Dict[int, int]:
-        raise NotImplementedError
+        return self._can_swap(seq_group, Device.GPU, SequenceStatus.SWAPPED,
+                              num_lookahead_slots, self.watermark_blocks)
+
+    def swap_in(self,
+                sequence_group: SequenceGroup,
+                num_lookahead_slots: int = 0) -> Dict[int, int]:
+        blocks = self._get_blocks_for_swap(sequence_group,
+                                           SequenceStatus.SWAPPED,
+                                           num_lookahead_slots)
+        self.block_allocator.swap(blocks=blocks,
+                                  source_device=Device.CPU,
+                                  dest_device=Device.GPU)
+        # NOTE: Once the BlockManagerV1 implementation is deleted, we can
+        # move this get_and_reset_swaps call outside of swap_in/swap_out.
+        # Then the scheduler can make calls to get all swaps and all
+        # copy-on-writes for the batch.
+        mapping = self.block_allocator.get_and_reset_swaps()
+        block_number_mapping = {
+            cpu_block_id - self.num_total_gpu_blocks: gpu_block_id
+            for cpu_block_id, gpu_block_id in mapping.items()
+        }
+        return block_number_mapping
```

Comment:
this will break if the CpuGpu block allocator ever changes its mapping, e.g. CPU blocks in low range and GPU blocks in high range. can we decouple them?

Reply:
same story in `swap_out`

Reply:
Right. Pushed it down to the Block allocator to get the relative block id with specific allocator.

Reply:
`cpu_block_id - self.num_total_gpu_blocks` this still breaks if the mapping changes

Reply:
The code here is outdated, please refer to the newest version.

Reply:
The code here is outdated, please refer to the new version!